metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -336,10 +342,14 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
                 ", but got cluster ID " + request.clusterId());
         }
         int brokerId = request.brokerId();
+        List<ApiMessageAndVersion> records = new ArrayList<>();
         BrokerRegistration existing = brokerRegistrations.get(brokerId);
         if (version < 2 || existing == null || request.previousBrokerEpoch() != existing.epoch()) {
-            // TODO(KIP-966): Update the ELR if the broker has an unclean shutdown.
-            log.debug("Received an unclean shutdown request");
+            if (handleBrokerUncleanShutdownHelper == null) {
+                log.warn("No handleBrokerUncleanShutdownHelper provided");
```

Comment:
Related to the comment about the builder, we should do this validation when constructing the class so we can throw exceptions on startup if the handler was somehow not specified. 

To work around the circular dependency of ClusterControlManager and ReplicaManager, you can add something like this to QuorumController:

```
void handleUncleanBrokerShutdown(int brokerId, List<> records) {
  replicationControl.handleBrokerUncleanShutdown(brokerId, records);
}
```

and use that method in the builder of ClusterControlManager.

```
 setUncleanBrokerShutdownHandler(QuorumController.this::handleUncleanBrokerShutdown)
```

Reply:
Thanks for the advice! Updated.